I refreshed my data cleaning skills.

###### Import libraries and loading of data

In [8]:
import pandas as pd
import numpy as np
data = pd.read_csv('netflix_titles.csv')
data.head()

show_id     type  title           director  \
0      s1  TV Show     3%                NaN   
1      s2    Movie   7:19  Jorge Michel Grau   
2      s3    Movie  23:59       Gilbert Chan   
3      s4    Movie      9        Shane Acker   
4      s5    Movie     21     Robert Luketic   

                                                cast        country  \
0  João Miguel, Bianca Comparato, Michel Gomes, R...         Brazil   
1  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...         Mexico   
2  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...      Singapore   
3  Elijah Wood, John C. Reilly, Jennifer Connelly...  United States   
4  Jim Sturgess, Kevin Spacey, Kate Bosworth, Aar...  United States   

          date_added  release_year rating   duration  \
0    August 14, 2020          2020  TV-MA  4 Seasons   
1  December 23, 2016          2016  TV-MA     93 min   
2  December 20, 2018          2011      R     78 min   
3  November 16, 2017          2009  PG-13     80 min   
4    January 1, 2020          2008  PG-13    123 min   

                                           listed_in  \
0  International TV Shows, TV Dramas, TV Sci-Fi &...   
1                       Dramas, International Movies   
2                Horror Movies, International Movies   
3  Action & Adventure, Independent Movies, Sci-Fi...   
4                                             Dramas   

                                         description  
0  In a future where the elite inhabit an island ...  
1  After a devastating earthquake hits Mexico Cit...  
2  When an army recruit is found dead, his fellow...  
3  In a postapocalyptic world, rag-doll robots hi...  
4  A brilliant group of students become card-coun...

#### Check for missing data

In [6]:
for i in data.columns:
    null_rate = data[i].isna().sum()/len(data)*100
    if null_rate > 0:
        print("Column '{}' null rate: {}%".format(i,round(null_rate,2)))

Column 'director' null rate: 30.68%
Column 'cast' null rate: 9.22%
Column 'country' null rate: 6.51%
Column 'date_added' null rate: 0.13%
Column 'rating' null rate: 0.09%


###### 5 columns have missing values, with column Director almost 1/3 of data missing. We can keep all the columns that contains missing data, just make some tweaks here and there, like:
- replace missing countries with the most common country
- raplace missing data in 'Director' and 'Cast' with 'No information'

In [10]:
data['country'] = data['country'].fillna(data['country'].mode()[0])
data['cast'].replace(np.nan, 'No information', inplace=True)
data['director'].replace(np.nan, 'No information', inplace=True)

data.dropna(inplace=True)
data.drop_duplicates(inplace=True)

In [11]:
data.isnull().sum()

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64

In [12]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7770 entries, 0 to 7786
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       7770 non-null   object
 1   type          7770 non-null   object
 2   title         7770 non-null   object
 3   director      7770 non-null   object
 4   cast          7770 non-null   object
 5   country       7770 non-null   object
 6   date_added    7770 non-null   object
 7   release_year  7770 non-null   int64 
 8   rating        7770 non-null   object
 9   duration      7770 non-null   object
 10  listed_in     7770 non-null   object
 11  description   7770 non-null   object
dtypes: int64(1), object(11)
memory usage: 789.1+ KB


#### We have to change 'date_added' columns Dtype from 'object' to 'datetime' in order to perform time series based operation on the dates.

In [14]:
data['date_added'] = pd.to_datetime(data['date_added'])
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7770 entries, 0 to 7786
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       7770 non-null   object        
 1   type          7770 non-null   object        
 2   title         7770 non-null   object        
 3   director      7770 non-null   object        
 4   cast          7770 non-null   object        
 5   country       7770 non-null   object        
 6   date_added    7770 non-null   datetime64[ns]
 7   release_year  7770 non-null   int64         
 8   rating        7770 non-null   object        
 9   duration      7770 non-null   object        
 10  listed_in     7770 non-null   object        
 11  description   7770 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(10)
memory usage: 789.1+ KB


In [16]:
data.head(3)

show_id     type  title           director  \
0      s1  TV Show     3%     No information   
1      s2    Movie   7:19  Jorge Michel Grau   
2      s3    Movie  23:59       Gilbert Chan   

                                                cast    country date_added  \
0  João Miguel, Bianca Comparato, Michel Gomes, R...     Brazil 2020-08-14   
1  Demián Bichir, Héctor Bonilla, Oscar Serrano, ...     Mexico 2016-12-23   
2  Tedd Chan, Stella Chung, Henley Hii, Lawrence ...  Singapore 2018-12-20   

   release_year rating   duration  \
0          2020  TV-MA  4 Seasons   
1          2016  TV-MA     93 min   
2          2011      R     78 min   

                                           listed_in  \
0  International TV Shows, TV Dramas, TV Sci-Fi &...   
1                       Dramas, International Movies   
2                Horror Movies, International Movies   

                                         description  
0  In a future where the elite inhabit an island ...  
1  After a devastating earthquake hits Mexico Cit...  
2  When an army recruit is found dead, his fellow...

#### Take another dataset

In [17]:
data2 = pd.read_csv('googleplaystore.csv')

#### Familiarising myself with a dataset

In [18]:
data2.columns

Index(['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type',
       'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver',
       'Android Ver'],
      dtype='object')

In [19]:
data2.head()

App        Category  Rating  \
0     Photo Editor & Candy Camera & Grid & ScrapBook  ART_AND_DESIGN     4.1   
1                                Coloring book moana  ART_AND_DESIGN     3.9   
2  U Launcher Lite – FREE Live Cool Themes, Hide ...  ART_AND_DESIGN     4.7   
3                              Sketch - Draw & Paint  ART_AND_DESIGN     4.5   
4              Pixel Draw - Number Art Coloring Book  ART_AND_DESIGN     4.3   

  Reviews  Size     Installs  Type Price Content Rating  \
0     159   19M      10,000+  Free     0       Everyone   
1     967   14M     500,000+  Free     0       Everyone   
2   87510  8.7M   5,000,000+  Free     0       Everyone   
3  215644   25M  50,000,000+  Free     0           Teen   
4     967  2.8M     100,000+  Free     0       Everyone   

                      Genres      Last Updated         Current Ver  \
0               Art & Design   January 7, 2018               1.0.0   
1  Art & Design;Pretend Play  January 15, 2018               2.0.0   
2               Art & Design    August 1, 2018               1.2.4   
3               Art & Design      June 8, 2018  Varies with device   
4    Art & Design;Creativity     June 20, 2018                 1.1   

    Android Ver  
0  4.0.3 and up  
1  4.0.3 and up  
2  4.0.3 and up  
3    4.2 and up  
4    4.4 and up

In [20]:
data2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10841 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   App             10841 non-null  object 
 1   Category        10841 non-null  object 
 2   Rating          9367 non-null   float64
 3   Reviews         10841 non-null  object 
 4   Size            10841 non-null  object 
 5   Installs        10841 non-null  object 
 6   Type            10840 non-null  object 
 7   Price           10841 non-null  object 
 8   Content Rating  10840 non-null  object 
 9   Genres          10841 non-null  object 
 10  Last Updated    10841 non-null  object 
 11  Current Ver     10833 non-null  object 
 12  Android Ver     10838 non-null  object 
dtypes: float64(1), object(12)
memory usage: 1.1+ MB


###### This dataset has 10841 rows and 13 columns. 

In [21]:
data2.isnull().sum()

App                  0
Category             0
Rating            1474
Reviews              0
Size                 0
Installs             0
Type                 1
Price                0
Content Rating       1
Genres               0
Last Updated         0
Current Ver          8
Android Ver          3
dtype: int64

There are **1474** missing values in column 'Rating', **1** in 'Type', **1** in 'Content Rating', **8** in 'Current Ver' and **3** more missing values in column 'Android Ver'.

In [22]:
(data2.isnull().sum()/data2.shape[0])*100

App                0.000000
Category           0.000000
Rating            13.596532
Reviews            0.000000
Size               0.000000
Installs           0.000000
Type               0.009224
Price              0.000000
Content Rating     0.009224
Genres             0.000000
Last Updated       0.000000
Current Ver        0.073794
Android Ver        0.027673
dtype: float64

###### Table above shows missing values expressed as a percentage of total data. Column 'Rating' contains the most missed values (13,59%).

Lets say we want to fill all of those missing values in column 'Rating' with average rating **'2.5'** so it looks all nice and neatly.

In [23]:
data2['Rating'].fillna(value='2.5', inplace=True)

**Try again** to count the sum of all of the missing values.

In [27]:
data2.isnull().sum()

App               0
Category          0
Rating            0
Reviews           0
Size              0
Installs          0
Type              1
Price             0
Content Rating    1
Genres            0
Last Updated      0
Current Ver       8
Android Ver       3
dtype: int64

Now all **missing** values in column **'Rating'** has changed to **'2.5'**.

###### Remove dublicate values

In [29]:
(data2.duplicated().value_counts()/data2.shape[0])*100

False    95.544691
True      4.455309
dtype: float64

###### 4,45% of all values is dublicates - drop them.

In [30]:
data2.drop_duplicates(inplace=True)

data2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10358 entries, 0 to 10840
Data columns (total 13 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   App             10358 non-null  object
 1   Category        10358 non-null  object
 2   Rating          10358 non-null  object
 3   Reviews         10358 non-null  object
 4   Size            10358 non-null  object
 5   Installs        10358 non-null  object
 6   Type            10357 non-null  object
 7   Price           10358 non-null  object
 8   Content Rating  10357 non-null  object
 9   Genres          10358 non-null  object
 10  Last Updated    10358 non-null  object
 11  Current Ver     10350 non-null  object
 12  Android Ver     10355 non-null  object
dtypes: object(13)
memory usage: 1.1+ MB


In [31]:
(data2.duplicated().value_counts()/data2.shape[0])*100

False    100.0
dtype: float64

###### Now all dublicates are gone.

### Let's explore and clean the last dataset.

In [34]:
data3 = pd.read_csv('players_15.csv')
data3.head()

sofifa_id                                         player_url  \
0     158023  https://sofifa.com/player/158023/lionel-messi/...   
1      20801  https://sofifa.com/player/20801/c-ronaldo-dos-...   
2       9014  https://sofifa.com/player/9014/arjen-robben/15...   
3      41236  https://sofifa.com/player/41236/zlatan-ibrahim...   
4     167495  https://sofifa.com/player/167495/manuel-neuer/...   

          short_name                            long_name  age         dob  \
0           L. Messi       Lionel Andrés Messi Cuccittini   27  1987-06-24   
1  Cristiano Ronaldo  Cristiano Ronaldo dos Santos Aveiro   29  1985-02-05   
2          A. Robben                         Arjen Robben   30  1984-01-23   
3     Z. Ibrahimović                   Zlatan Ibrahimović   32  1981-10-03   
4           M. Neuer                         Manuel Neuer   28  1986-03-27   

   height_cm  weight_kg  nationality                 club  ...   lwb   ldm  \
0        169         67    Argentina         FC Barcelona  ...  62+3  62+3   
1        185         80     Portugal          Real Madrid  ...  63+3  63+3   
2        180         80  Netherlands    FC Bayern München  ...  64+3  64+3   
3        195         95       Sweden  Paris Saint-Germain  ...  61+3  65+3   
4        193         92      Germany    FC Bayern München  ...   NaN   NaN   

    cdm   rdm   rwb    lb   lcb    cb   rcb    rb  
0  62+3  62+3  62+3  54+3  45+3  45+3  45+3  54+3  
1  63+3  63+3  63+3  57+3  52+3  52+3  52+3  57+3  
2  64+3  64+3  64+3  55+3  46+3  46+3  46+3  55+3  
3  65+3  65+3  61+3  56+3  55+3  55+3  55+3  56+3  
4   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN  

[5 rows x 104 columns]

##### Wow, 104 columns. Let's print them out to find out what they are.

In [35]:
list(data3.columns)

['sofifa_id',
 'player_url',
 'short_name',
 'long_name',
 'age',
 'dob',
 'height_cm',
 'weight_kg',
 'nationality',
 'club',
 'overall',
 'potential',
 'value_eur',
 'wage_eur',
 'player_positions',
 'preferred_foot',
 'international_reputation',
 'weak_foot',
 'skill_moves',
 'work_rate',
 'body_type',
 'real_face',
 'release_clause_eur',
 'player_tags',
 'team_position',
 'team_jersey_number',
 'loaned_from',
 'joined',
 'contract_valid_until',
 'nation_position',
 'nation_jersey_number',
 'pace',
 'shooting',
 'passing',
 'dribbling',
 'defending',
 'physic',
 'gk_diving',
 'gk_handling',
 'gk_kicking',
 'gk_reflexes',
 'gk_speed',
 'gk_positioning',
 'player_traits',
 'attacking_crossing',
 'attacking_finishing',
 'attacking_heading_accuracy',
 'attacking_short_passing',
 'attacking_volleys',
 'skill_dribbling',
 'skill_curve',
 'skill_fk_accuracy',
 'skill_long_passing',
 'skill_ball_control',
 'movement_acceleration',
 'movement_sprint_speed',
 'movement_agility',
 'movement_re

In [36]:
data3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 15465 entries, 0 to 15464
Columns: 104 entries, sofifa_id to rb
dtypes: float64(17), int64(11), object(76)
memory usage: 12.3+ MB


##### I would say we can remove some of unnecessary columns

In [37]:
unnecessary_columns = ['sofifa_id', 'player_url', 'long_name', 'dob', 'loaned_from']
data3 = data3.drop(unnecessary_columns, axis = 1)

###### Find the missing values of different columns.

In [41]:
null_data = data3.isna().sum().sort_values(ascending=False)
null_data = null_data.reset_index(drop = False)
null_data = null_data.rename(columns={"index": "Columns", 0:"Value"})
null_data['proportion'] = (null_data['Value']/len(data3))*100
null_data.head()

Columns  Value  proportion
0   mentality_composure  15465  100.000000
1    release_clause_eur  15465  100.000000
2       nation_position  14409   93.171678
3  nation_jersey_number  14409   93.171678
4           player_tags  14271   92.279340

###### Look at dataset after removing unnecessary columns

In [39]:
data3.shape

(15465, 99)

#### It's always fun to glance at statistical description of data.

In [40]:
data3.describe()

age     height_cm     weight_kg       overall     potential  \
count  15465.000000  15465.000000  15465.000000  15465.000000  15465.000000   
mean      24.763272    181.093631     75.482703     63.948594     68.483091   
std        4.624565      6.635182      6.907243      7.208610      6.611708   
min       16.000000    155.000000     50.000000     40.000000     40.000000   
25%       21.000000    176.000000     70.000000     59.000000     64.000000   
50%       24.000000    181.000000     75.000000     64.000000     68.000000   
75%       28.000000    186.000000     80.000000     69.000000     73.000000   
max       44.000000    203.000000    110.000000     93.000000     95.000000   

       value_eur  wage_eur  international_reputation     weak_foot  \
count    15465.0   15465.0              15465.000000  15465.000000   
mean         0.0       0.0                  1.126350      2.932363   
std          0.0       0.0                  0.401362      0.652270   
min          0.0       0.0                  1.000000      1.000000   
25%          0.0       0.0                  1.000000      3.000000   
50%          0.0       0.0                  1.000000      3.000000   
75%          0.0       0.0                  1.000000      3.000000   
max          0.0       0.0                  5.000000      5.000000   

        skill_moves  ...     dribbling     defending        physic  \
count  15465.000000  ...  13762.000000  13762.000000  13762.000000   
mean       2.267055  ...     60.617861     50.000799     64.602456   
std        0.719035  ...     10.800142     15.850317      9.698882   
min        1.000000  ...     25.000000     21.000000     28.000000   
25%        2.000000  ...     54.000000     33.000000     58.000000   
50%        2.000000  ...     62.000000     53.000000     66.000000   
75%        3.000000  ...     68.000000     63.000000     72.000000   
max        5.000000  ...     96.000000     90.000000     91.000000   

         gk_diving  gk_handling   gk_kicking  gk_reflexes     gk_speed  \
count  1703.000000  1703.000000  1703.000000  1703.000000  1703.000000   
mean     64.709336    60.823840    59.455079    65.955373    43.900763   
std       7.855747     8.510626     8.689929     8.374343     8.440858   
min      35.000000    36.000000    21.000000    31.000000    20.000000   
25%      59.000000    55.000000    53.000000    60.000000    39.000000   
50%      64.000000    61.000000    59.000000    65.000000    44.000000   
75%      70.000000    67.000000    65.000000    72.000000    49.000000   
max      88.000000    87.000000    92.000000    90.000000    66.000000   

       gk_positioning  mentality_composure  
count     1703.000000                  0.0  
mean        61.230182                  NaN  
std          9.188257                  NaN  
min         38.000000                  NaN  
25%         54.000000                  NaN  
50%         61.000000                  NaN  
75%         68.000000                  NaN  
max         90.000000                  NaN  

[8 rows x 27 columns]

#### We see that, for example, the average height of player is 181 cm (mean) or that the oldest player in data had 44 years.